### PR TITLE
#157720834 Completes user sign up endpoint

### PR DESCRIPTION
**Pull request by chukwuemekachm**
#### What does this PR do?
- Completes user signup endpoint
#### Description of Task to be completed?
- Write tests for endpoints
- Create POST route to api/v1/auth/signup
- Validate user inputs
- Persist data
- Test using Postman
#### How should this be manually tested?
- Clone the repo
- Startup the command line
- run ``` npm install ```
- run ```npm run start:dev ```
- Test endpoint using Postman
#### Any background context you want to provide?
- You should add a ```.env``` file to the project root
- In the ```.env``` file, specify the ```DATABASE_URL``` as an env_variable
#### What are the relevant pivotal tracker stories?
- #157720834
#### Screenshots (if appropriate)
screenshot 159
screenshot 158

#### Questions:
- None

### CODE REVIEW

File: Server/controllers/userController.js [hunk 1]
```diff
@@ -1,116 +1,51 @@
...
-      });
-    }
-    return res.status(200).json({
-      status: 'success',
-      code: 200,
-      message: 'Request found for the user',
-      data: { request },
+  static signUp(req, res, next) {
+    const client = new Client({
+      connectionString : connectionString,
```
Comment (paradisekelechi) [inline Server/controllers/userController.js]: Eslint should have complained thus `Expected property shorthand. (object-shorthand)` here. Are you sure your eslint is properly configured? @chukwuemekachm

File: Server/controllers/userController.js [hunk 2]
```diff
@@ -1,116 +1,51 @@
...
-      req.body.request.description,
-    );
-    Request.createRequest(request);
-    return res.status(201).json({
-      status: 'success',
-      code: 201,
-      message: 'Request created successfully',
-      data: { request },
+    client.connect();
+    const query = {
```
Comment (paradisekelechi) [inline Server/controllers/userController.js]: Kindly rename this accordingly. `queryString` is a better name

File: Server/controllers/userController.js [hunk 3]
```diff
@@ -1,116 +1,51 @@
...
+    text: 'INSERT INTO users(firstname, lastname, email, password) VALUES($1, $2, $3, $4) RETURNING id, firstname, lastname, email, createdat;',
+    values: [req.body.user.firstname, req.body.user.lastname, req.body.user.email, req.body.user.password]
```
Comment (paradisekelechi) [inline Server/controllers/userController.js]: Use object destructuring here

File: Server/controllers/userController.js (hunk 2 above)
Comment (chukwuemekachm) [reply Server/controllers/userController.js]: Ok I will

File: Server/controllers/userController.js [hunk 4]
```diff
@@ -1,116 +1,51 @@
...
+    };
+    client.query(query, (error, result) => {
+      client.end();
+      if (error) {
+        return res.status(400)
```
Comment (paradisekelechi) [inline Server/controllers/userController.js]: Good implementation and understanding of the architecture

File: Server/controllers/userController.js [hunk 5]
```diff
@@ -1,116 +1,51 @@
...
+            status: 'fail',
+            code: 400,
+            message: 'User sign up failed',
+          });
+      } else {
+        return res.status(201)
+          .json({
+            status: 'success',
+            code: 201,
+            data: { fullname: `${req.body.user.firstname} ${req.body.user.lastname}`, email: req.body.user.email },
```
Comment (paradisekelechi) [inline Server/controllers/userController.js]: Using object-destructuring would help you

File: Server/controllers/userController.js (hunk 4 above)
Comment (chukwuemekachm) [reply Server/controllers/userController.js]: Thanks Boss

File: Server/middlewares/check.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Checks if a user with the email already exists
+ *
+ * @param {object} req - The request object
+ * @param {object} res - The response object
+ * @param {object} done - The next middleware to be called
+ */
+const signupCheck = (req, res, done) => {
+  const query = {
```
Comment (paradisekelechi) [inline Server/middlewares/check.js]: Rename this accordingly

File: Server/middlewares/validation.js [hunk 7]
```diff
@@ -1,98 +1,70 @@
...
+  if (!req.body.firstname || typeof req.body.firstname !== 'string' || !/^[a-zA-Z]{1,}$/.test(req.body.firstname)) {
     return res.status(400).json({
       status: 'error',
       code: 400,
-      message: 'title is required or invalid',
+      message: 'firstname is required or invalid',
     });
   }
-  if (!req.body.type || typeof req.body.type !== 'string' || !/^[a-zA-Z]{1,}$/.test(req.body.type)) {
+  if (!req.body.lastname || typeof req.body.lastname !== 'string' || !/^[a-zA-Z]{1,}$/.test(req.body.lastname)) {
```
Comment (paradisekelechi) [inline Server/middlewares/validation.js]: Use destructuring here

File: Server/controllers/userController.js [hunk 8]
```diff
@@ -1,116 +1,53 @@
...
-    Request.updateRequest(parseInt(req.params.id, 10), request);
-    return res.status(200).json({
-      status: 'success',
-      code: 200,
-      message: 'Request updated successfully',
-      data: { request },
+    client.connect();
+    const {
+      firstname, lastname, email, password,
+    } = req.body.user;
```
Comment (paradisekelechi) [inline Server/controllers/userController.js:26]: Good destructuring

File: Server/models/db.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+const migrateUserDb = 'DROP TABLE IF EXISTS users CASCADE; CREATE TABLE users (id SERIAL NOT NULL PRIMARY KEY, firstname VARCHAR(20) NOT NULL, lastname VARCHAR(20) NOT NULL, email VARCHAR(20) NOT NULL UNIQUE, admin BOOLEAN NOT NULL DEFAULT FALSE, password VARCHAR(100) NOT NULL, createdat TIMESTAMP NOT NULL DEFAULT CURRENT_DATE, updatedat TIMESTAMP NOT NULL DEFAULT CURRENT_DATE );';
+const seedSuperUser = 'INSERT INTO users (firstname,lastname,email,admin,password) VALUES (\'Chima\',\'Emeka\',\'emecus10@gmail.com\',true,\'123456\');';
+const Query = `${migrateUserDb} ${seedSuperUser}`;
```
Comment (paradisekelechi) [inline Server/models/db.js]: The naming of `Query` is not camelCase

File: package.json [hunk 10]
```diff
@@ -25,15 +24,16 @@
...
-        "db-migrate": "^0.11.1",
-        "db-migrate-pg": "^0.4.0",
         "errorhandler": "^1.5.0",
         "express": "^4.16.3",
         "joi": "^13.3.0",
         "jsonwebtoken": "^8.2.1",
         "morgan": "^1.9.0",
-        "pg": "^7.4.3"
+        "pg": "^7.0.3",
+        "pg-promise": "^8.4.4",
```
Comment (paradisekelechi) [inline package.json]: Remove `pg-promise` since you are not using it

**Pull request merged**